File: internal/application/usecase/user/delete.go
package user

import (
	"context"

	"go-clean-template/internal/domain"
)

type DeleteUserInput struct {
	ID string
}

type DeleteUserUseCase interface {
	Execute(ctx context.Context, input *DeleteUserInput) error
}

type deleteUserUseCase struct {
	userRepo domain.UserRepository
}

func NewDeleteUserUseCase(userRepo domain.UserRepository) DeleteUserUseCase {
	return &deleteUserUseCase{userRepo: userRepo}
}

func (uc *deleteUserUseCase) Execute(ctx context.Context, input *DeleteUserInput) error {
	if _, err := uc.userRepo.FindByID(ctx, input.ID); err != nil {
		return err
	}
	return uc.userRepo.Delete(ctx, input.ID)
}

File: src/application/usecase/user/delete.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::UserRepository;

pub struct DeleteUserInput {
    pub id: String,
}

#[async_trait]
pub trait DeleteUserUseCase: Send + Sync {
    async fn execute(&self, input: DeleteUserInput) -> Result<(), AppError>;
}

pub struct DeleteUserUseCaseImpl {
    user_repo: Arc<dyn UserRepository>,
}

impl DeleteUserUseCaseImpl {
    pub fn new(user_repo: Arc<dyn UserRepository>) -> Self {
        DeleteUserUseCaseImpl { user_repo }
    }
}

#[async_trait]
impl DeleteUserUseCase for DeleteUserUseCaseImpl {
    async fn execute(&self, input: DeleteUserInput) -> Result<(), AppError> {
        self.user_repo.find_by_id(&input.id).await?;
        self.user_repo.delete(&input.id).await?;
        Ok(())
    }
}
